### ISSUE TITLE
Fix encapsulation order

### ISSUE THREAD

**Issue by mbroecheler** [thinkaurelius/titan#971]
The nutshell explanation is that https://github.com/thinkaurelius/titan/commit/278d8b988058183ed032b03959771864f6b37a0c switched the encapsulation order of the Metrics and TTL StoreManager wrappers, but the Metrics wrapper didn't implement an interface expected by the TTL wrapper (the CustomizeStoreKCVSManager interface with the openDatabase(name, ttl) overload).  HBase still implements the interface expected by the TTL wrapper, but it's effectively hidden by the Metrics wrapper now sitting between them in the encapsulation order.  This leads to an exception on HBase with Metrics enabled as Backend tries to open a KCVSLogManager.

I tried to apply the simplest fix possible in https://github.com/thinkaurelius/titan/commit/d78a0801de473a1943061ef616d0252b55b98f66.  Mixing the decorator pattern and instanceof is likely to bite us again, and it might be worthwhile to try an alternate approach.  But I don't want to go nuts with refactoring right before a release, so I applied a low-effort fix.

**Comment by dalaro** [comment]
I pushed a branch containing a potential fix for this (issue_971), but it adds nearly as many lines as it removes.  This may not even be worth doing.  I also moved the milestone to titan09, since the approach I took added an `openDatabase` overload signature to the KCVS interface.

**Comment by mbroecheler** [comment]
If it doesn't simplify the code, then I'd say we drop it.

**Comment by dalaro** [comment]
I reworked this and pushed new commits on the issue_971 branch.  I'm not going to merge it until all TTL tests are passing, and there's at least one preexisting failure that I need to resolve first (#1036).

Anyway, in the branch right now, KCVS defines two openDatabase methods:
- `openDatabase(String name, StoreMetaData.Container options)` with no default impl
- `openDatabase(String name)` with a default impl that calls the preceding method with an empty options container

The `StoreMetaData.Container` class uses the typesafe heterogeneous container pattern.  This means store impls don't have to cast TTL's associated object to an Integer anymore.  The only StoreMetaData key is TTL, though we could add new ones in the future.  Like EntryMetaData, StoreMetaData is not open to third-party extension, since core has to support any new members for them to be useful.

Backends with store-level TTL are now usable directly, without wrapping in a TTLKCVSManager.  TTLKCVSManager is now only used for converting a backend with cell-level TTL into one with store-level TTL, e.g. KCVSLog on Cassandra.  I also fixed the old typo in TTLKCVSManager's name (V and C were transposed).

The diffstat for the issue_971 adds more than it deletes (mostly StoreMetaData's typesafe heterogeneous container):

```
 17 files changed, 263 insertions(+), 174 deletions(-)
```

The best argument against this change is that it adds a new method to the KCVS interface.  Here's a recap of why this may be worthwhile.

The old TTL approach avoids changing KCVS in favor of introducing a separate interface, `CustomizeKCVS`, that defines a single method: `openDatabase(String name, int ttlSeconds)`.  Backends with store-level TTL had to implement `CustomizeKCVS`.  That approach was fine, except that wrapping our stores in layers of decorators meant that the `CustomizeKCVS` implementation could be inadvertently masked by seemingly innocuous statement reordering.

The approach in the issue_971 branch should avoid this problem.  Casting or `instanceof` is no longer necessary to either detect or use TTL.  The new approach uses `StoreFeatures` to detect TTL support and the new `StoreMetaData.Container` parameter to pass a value when TTL support is detected.
